### ISSUE TITLE
RFC: Distributed Deployment Configuration

### ISSUE THREAD

**Issue by mapl** [TwinProduction/gatus#64]
Assume you have Gatus deployed in several security subnets (or zones) to monitor individual services because one single Gatus instance is not able to reach those services (administratively prohibited due to firewall rules, etc.)

But you want one main Gatus instance which is capable of retrieving health information of services from all those other Gatus instances to display them in one unified Gatus Dashboard.

Can we start a short discussion on this?

PS. Thank you for this great project!

**Comment by TwinProduction** [comment]
Hello @mapl.

This is quite an edge case, so honestly, I'm not sure I'll be personally working on this unless more people show interest for this feature (based on the number of 👍 on the issue).

The easiest way to implement this that I can think of is by having a way to allow pushing data into Gatus, as opposed to the current behavior, which only allows retrieving data from Gatus.
By leveraging this as well as a configuration that allows specifying whether a Gatus instance is the primary instance or a secondary instance, the latter required to specify the endpoint of the primary instance, it would be possible to have a "global" dashboard and multiple Gatus instances configured independently. 

Fortunately, the easiest way is also the most convenient one, because the other ones would likely involve persistence.

There's an even easier solution, but it assumes that the users accessing the dashboard has access to all "security subnets/zones", which I'm not sure is the case based on your explanation. The only work required for to send a request from the Gatus' dashboard frontend to each backends and merge the statuses.

**Comment by mapl** [comment]
I made a quick diagram about a simple distributed Gatus Deployment where the main Gatus Instance just pulls data from remote Gatus instances.

The Data from each remote Gatus instance is just embedded into the Main Gatus Dashboard.
 
![gatus](https://user-images.githubusercontent.com/9784732/105615093-37751c00-5dce-11eb-8d88-05059602d565.png)

**Comment by sefaphlvn** [comment]
I think this is a good idea. I would like it if there was

**Comment by TwinProduction** [comment]
That looks good.

@mapl What do you think would be the appropriate behavior when there are overlapping service names?

Also, how about we do the opposite: the remote Gatus instances push their data to the main Gatus instance? 
I think that would allow a lot more flexibility, especially if, for instance, one of the remote Gatus instances is running in an environment completely inaccessible from the main Gatus instance (i.e. locally).

Of course, this would require a layer of security, but I built something oddly relevant to this specific use case: https://github.com/TwinProduction/g8

We'd also need to add something that periodically cleans up services that haven't been refreshed in a long time (i.e. in case a remote instance is taken offline, we don't want to keep the outdated service health checks on the dashboard forever).

**Comment by mapl** [comment]
### Overlapping service names

Hmm... not sure. Is this actually supposed to be a problem? A UUID would solve the most of the issues I think.
____
### Clean up services

A steady Health Status of the remote Gatus instance would be cool, like the last reached Status, just to know if its still alive.

___

There is and always been an ongoing debate about Pulling VS Pushing when it comes to monitoring.

For example, if you have 100 remote Gatus instances and each of those instances would push data to the main Gatus instance, the main instance is quickly subject to an overload of metric data.

I am wondering if it is enough to just proxy through data from the remote instances to the main instance, or is caching needed?

The goal should always be a simple as possible design.

https://dave.cheney.net/2019/07/09/clear-is-better-than-clever


### Security 

An API Token is commonly used to restrict permissions. What's your opinion on this?

___



Anyway, a very goof reference is Prometheus and why the devs decide to go with pulling rather than pushing.

However, that's not a one way road, so it depends on the scenario. For the most part Pulling is the better choice.

https://prometheus.io/blog/2016/07/23/pull-does-not-scale-or-does-it/

https://prometheus.io/docs/introduction/faq/#why-do-you-pull-rather-than-push

**Comment by TwinProduction** [comment]
Fair enough, though if we're pulling, then there's no need for a token since the endpoint needs to be public for the dashboard to be shown.

Sounds good!

**Comment by mapl** [comment]
If you can ensure that the main Gatus instance can reach its remote instances via http(s) to fetch the json data, it should be fine

What do you think of this config idea snippet to be deployed on the main instance? I thought an Api key could be handy but not required as you mentioned.

```
remote-services:
  - name: Gatus Remote Instance 1	
    url: http://10.10.10.10/api/v1/statuses
    api-key: XXXXXXXXXXXXXX
    interval: 10s
  - name: Gatus Remote Instance 2 	
    url: http://10.10.10.20/api/v1/statuses
    api-key: XXXXXXXXXXXXXX
    interval: 5s
  - name: Gatus Remote Instance 3
    url: http://10.10.10.30/api/v1/statuses
    api-key: XXXXXXXXXXXXXX
    interval: 10s
```

**Comment by TwinProduction** [comment]
I was thinking something along the lines of 
```yaml
remote:
  instances:
    - name: "gatus-internal"
      url: "http://10.10.10.10/api/v1/statuses"
      interval: 30s
```
because in the future, we may have to add other parameters specific to the `remote` configuration, such as:
- `strategy: merge`: merge all statuses fetched from the remote instances with existing, overlapping service names, or create new statuses if they already exist
![image](https://user-images.githubusercontent.com/15699766/106674764-e27d9680-6581-11eb-8b93-ad2d2395c05f.png)


- `strategy: prefix-service`: Prepend each retrieved service names with the configured name of the remote instance
![image](https://user-images.githubusercontent.com/15699766/106674827-f628fd00-6581-11eb-933b-d4a696c1de11.png)


- `strategy: prefix-group`: Prepend each retrieved group names with the configured name of the remote instance
![image](https://user-images.githubusercontent.com/15699766/106674838-faedb100-6581-11eb-9bf0-be1f6bb8dfb9.png)

Anyways, this isn't for right now, but I still think it's good to make the configuration as extensible as possible to prevent future breaking changes.

**Comment by mapl** [comment]
Absolutely, the config should be as extensible as possible and future proof where possible. Good point.
Your concerns about possible service names clashes when the main and remote instances have the same names could be an actual issue if you put them in one flat view. 
I think, if every remote instance is moved into its own, lets says "container" box, then you could clearly derive where its origin is.
So if you have duplicate names in this case, it wouldn't matter as they are displayed in their own container, in a way you added this nice grouping feature.
Additionally, you could also check the health state of the entire remote instance, about last time reached for instance, latency, etc.

**Comment by TwinProduction** [comment]
I've been thinking a bit more about the implementation for this, and one issue I can think of is how to handle alerting.

Assuming that the purpose is to monitor internal applications within a network that isn't publicly accessible by other users, but is accessible by a "remote" Gatus instance, how would alerting for a "remote" Gatus instance be handled, assuming that the alerting configuration as well as the individual alerts for each services are not exposed through Gatus' main endpoint (`/api/v1/statuses`)?

I think it's safe to assume that each remote instances are expected to deal with their own alerts and that the only difference between the remote instance(s) and the main instance is that the main instance's dashboard must display the statuses from the remote instances as well.

All in all, I don't think this is a problem, but felt like specifying this through a comment would be worthwhile for the sake of traceability.

**Comment by mcolladoio** [comment]
Maybe would be a good idea to have different systems?
One implemented with a check-in system, and another pulling data from intranet networks. 

Check-in system: Could give the control to developers to implement maybe some logic into the check (for example, in our systems we use healthchecks to run certain tests and if everything goes well, we send the ok to the healthchecks api endpoint, so we know that we are not only receiving that the server responds and is alive, plus adding complementary logic to the checks.  For example, knowing that in our databases we're receiving data or our auth service is working properly, for example)

* Statping: https://github.com/statping/statping
* Healthchecks: https://github.com/healthchecks/healthchecks

About the alerting on that mode of check-in, it is necessary to setup the expectation in time when the check-in must be in, and if is delayed, starts alerting.

**Comment by TwinProduction** [comment]
Haven't really had the time to work on this, unfortunately, but #124 may make implementing this much easier if we were to leverage a global database.

Each individual Gatus would be in charge of alerting for their respective services, _but_ the data would be retrieved from a global database that they all share (though this could be made configurable, in that the instance could choose to retrieve only the services it's monitoring, or all services present in the database)

There's obviously a few things that would need some thinking, like how to detect when one of the Gatus' instance configuration no longer has a given service (because it was deleted) so that we can automatically delete them.

Currently, since there's only one instance, there's no problem, but in a distributed setting, that won't work without a consensus of some sort, or maybe a table with a column to differentiate each individual Gatus instance as well as all the services registered under that instance could suffice?
e.g. 
- `gatus-1` has `service-a` and `service-b`
- `gatus-2` has `service-c` and `service-d`
- When `gatus-2`'s configuration is modified to remove `service-d`, `gatus-2` would update the table to remove `service-d` because according to the table, `gatus-2` was previously assigned to `service-d` but `service-d` is no longer in the configuration

**Comment by fatihbm** [comment]
There are good ideas in here. I hope this feature release as soon as possible. I'm looking forward to it 🙂

**Comment by TwinProduction** [comment]
With #136 merged, #124 is not that far off.

Note to self: Will probably need to add a parameter to control https://github.com/TwinProduction/gatus/blob/acb6757dc800b43b5a24e1fbe0ebf9f64b42df4f/storage/store/store.go#L25-L28

**Comment by TwinProduction** [comment]
Just to add on https://github.com/TwinProduction/gatus/issues/64#issuecomment-861882360 and https://github.com/TwinProduction/gatus/issues/64#issuecomment-896402326:

After giving it some thought, this is much easier than I initially anticipated.

The easiest, most barebone implementation I can think of is the following:
- [ ] #124 (easy because of #136)
- [ ] Add config parameter `distributed.enabled`. 
  - When set to `true`, three things happen:
    - `storage.type` must be set to `postgres`, or the application fails to start
    - Gatus doesn't require that at least one service is configured. This is for the scenario where you have a "public" instance in the DMZ whose only role is to let users view the dashboard -- it does not monitor any services its own, only pulls them from Postgres (which SHOULD be getting populated by the "private" Gatus instances).
    - `DeleteAllServiceStatusesNotInKeys` is never called. If a service is removed from one of the Gatus instances, it must be deleted from the database manually (keep in mind that this is a barebone implementation/MVP. Support for automatically cleaning up could be added later)

And that's it!
Even by a conservative estimate, this is less than a week of work. I just don't know when I'll find the time to work on it.

---

Just thinking about the possibilities that this could bring hypes me up quite a bit.

Consider the following: You're provisioning a fleet of clusters in different Cloud providers/accounts, and each of them have their own Gatus instance -- all of which are configured to monitor their respective "private" cluster, but they're pushing the data in a global Postgres database. To wrap everything up, you have a single "global" Gatus instance which doesn't monitor anything, but it's publicly available, and it exposes the data from each individual cluster. 

Anyways, I digress.

**Comment by appleboy** [comment]
I propose one solution: **server** and **agent**.

1. agent makes a new request to the server
2. server response what endpoint has not been handled.
3. agent starts the watchdog to handle the endpoint.

![gatus01](https://user-images.githubusercontent.com/21979/181446807-2a950585-a943-4927-b0d5-5eb5623eb063.png)

**Comment by TwiN** [comment]
So I just had a very simple idea for a temporary solution, and I just had to give it a shot (see #307).

I still think that the best approach to this problem will be by leveraging a shared database, but for now, this might do the trick for some of you.

Basically, all it does it retrieve the endpoint statuses from another remote host before returning the endpoint statuses. You may specify an endpoint prefix to prefix the name of all endpoints coming from a given remote instance with a string.

```yaml
remote:
  instances:
    - endpoint-prefix: "myremoteinstance-"
      url: "https://status.example.org/api/v1/endpoints/statuses"
```

Note that I haven't documented the feature yet, because it's experimental and it may be removed and/or updated.

Anyways, I'd love if some of you could give it a try it and let me know how it works.

**Comment by TwiN** [comment]
One of the issues with #307 is that clicking on the individual endpoint on the UI does not work. In other word, the page for viewing individual endpoints does not work if said endpoint comes from a remote Gatus instance.

I'm going to release this with v4.1.0, but I'm strongly considering getting rid of that implementation, unless somebody is actually using it and finds it helpful.

**Comment by nzsambo** [comment]
Thanks @TwiN for your efforts here, even if you're not happy with the implementation, it's a great start!

I would use this to monitor the same group of services from multiple locations, rather than just bypass security. I get that throwing an instance behind a NAT firewall is definitely great, even better is being able to see metrics from different perspectives of the Internet. My old smokeping installs highlight issues that are only evident from certain locations, perhaps caused by a particular ISP.

Monitoring the same set of targets from all offices of an organisation could also be an advantage.

The original problem this commit is trying to solve was secured subnets, and this implementation would require port forwarding or pinholes as you mentioned in a very early comment. On the whole push-pull argument my vote would definitely be push from the agents / remote clients. An api listening on 80/443 on the main instance with a simple static token I reckon. It gets around inbound firewalls. 

I have been meaning to try configuring remote gatus instances to use the database on the main instance. I would carefully configure the groups and services on each node. It wouldn't be ideal but it might work for me.

Anyway, I'll try this commit and report back.

Cheers!

Sam

**Comment by hypervtechnics** [comment]
I also have the same or a very similar use case as @nzsambo and I also agree on this being a push mechanism. 

I'd also recommend a static token per agent. The configuration should look something like this:

```yaml
# Doesn't matter whether array with name of agent or dictionary with name as key ;)
agents:
  - name: internet-pov
    token: abc123
    # Would contain a list of endpoints the agent is allowed to receive the configuration for (may also defined as a file on the agent) and update the status from their point of view:
    endpoints:
      - name: some-endpoint-defined-in-endpoints
        critical: true # If the agent is unavailable or reports the status for this endpoint as down/degraded this will mark the endpoint on the host as down instead of degraded
```

This would also require to add a new dimension for an endpoint check result to also contain the pov (agent or host). Maybe an option can be added to define the distribution behaviour with a **scope** (`host`: only the gatus host is allowed, `agent`: only the gatus agents are allowed may also be only one agent depending on the multiplier, all) and **multiplier/point of view mode**: (`single`: only one instance and will default to single if scope is host, `all`: all instances in scope)?

But please be aware of a version/api dependency from agent to host.

**Comment by xeruf** [comment]
I don't think this is an edge-case if you think bigger: I am considering a HA setup, but separate from the kubernetes cluster gatus should monitor to prevent entanglement. The proposed server-agent solutions also don't solve that, unless you can distribute the server, at which point you arrive at a design similar to kubernetes: server instance as control plane and agents as nodes.

I want an instance of gatus running inside our servers in Frankfurt, on the same machine as the services to monitor, and another one in the server on our office. Both should be visible in a single dashboard, but this dashboard needs to survive failure of any of the (two) nodes!
(I can add a third one if that's needed as typically in proper HA)

With HA like this, gatus would become a true enterprise-ready tool.

**Comment by hypervtechnics** [comment]
I think HA and a distributed deployment should be seen as two different things. Both have their challenges. How do you imagine a HA setup to work internally? Same database in the background? Syncing between the instances?

**Comment by xeruf** [comment]
Now that I revisit this, yes, both these concerns can be handled separately, and there is already quite an elaborate issue about HA: https://github.com/TwiN/gatus/issues/176

**Comment by hypervtechnics** [comment]
Some orientation from another project, which also contains some of the ideas I already mentioned: https://oss.oetiker.ch/smokeping/doc/smokeping_master_slave.en.html

**Comment by adamreed90** [comment]
We have multiple locations, a distrubuted testing agent is essential for us to ensure our services are available from all intranet locations and how they're performing. The inverse of most status pages (where one agent tests to various locations), we have been looking for the opposite forever! Would love to see this!

**Comment by TwiN** [comment]
FYI, [external endpoints](https://github.com/TwiN/gatus#external-endpoints) was implemented not long ago and can serve as a way to bypass connectivity challenges.

Long story short, rather than Gatus making the requests, you can now make the requests from within your own environments using whatever tool/application you have internally, and then push the results to Gatus.

**Comment by marrold** [comment]
> WARNING: This feature is a candidate for removal in future versions. Please comment on the issue above if you need this feature.

This feature seems useful to me, although the broken click on endpoint is unfortunate

**Comment by joryirving** [comment]
So I'm currently using external endpoints. I have one primary, externally available instance, as well as two local instances. It's polling those instances, and I'm wonder if it would be better to have those push instead.

Additionally, I wonder if pushing to the primary gatus would store the endpoints data in the DB, as primary is using psql, but the other two are using sqlite on emptyDir, so I lose all the data on container restart.

**Comment by bradbendy** [comment]
Not sure I should be posting this in here or a new PR. I have 3 remote-instances right now, everything works fine it seems. When a service on the remote-instance fails a check it reports it, all that works good.

I added a endpoint on the master node to do a HTTP check to the API and if that fails then alert me, figured it's the easiest way to see if the remote-instance is alive and well. In my testing if that endpoint fails OR ANY of the remote-instances go offline they disappear from the GUI and only the local endpoint appear. I would expect if one remote-instance is offline the others would still continue to work as nothing is dependent on that.

I looked at the disable-monitoring-lock but that does not change the behavior. The log shows "silently failed to retrieve endpoint statues" and never tries any others, just stops going to anything else in the config.

Is this the correct behavior and can't be a change or looks like a bug? I don't see how that would be useful since then you really have no idea what is done since every single endpoint disappears. Running the latest github download as of June 3rd 2024.

Thanks!

**Comment by muhlba91** [comment]
i have the same use case that there's one public instance but i have two secured networks where i want to display the service status on the public instance.
a pull as per the `remote` configuration doesn't work because the private instances are not reachable from outside the network and neither is `external-endpoints` helpful here, or at least i wasn't able to figure out how to configure a private gatus instance to push results to the external one.
i agree that some kind of server-agent model would be great where the agents are just pushing their results to the main/server instance.

**Comment by TheAnachronism** [comment]
> FYI, I implemented [external endpoints](https://github.com/TwiN/gatus#external-endpoints) not long ago and it can serve as a way to bypass connectivity challenges.
> 
> Long story short, rather than Gatus making the requests, you can now make the requests from within your own environments using whatever tool/application you have internally, and then push the results to Gatus.

Can Gatus itself act as that internal tool? Basically, just have two instances running and one pushes its results to the other.

**Comment by zubieta** [comment]
I was trying to just push some statuses from one gatus to another via external-enpoint and custom alerts. Unfortunately, it looks like it is not possible to trigger a request for every status check (since currently alerts only trigger on failures after some threshold). But I think it is worth checking if we could modify the alerting mechanism, since it might involve less extra implementation work.

**Comment by TwiN** [comment]
@zubieta that strategy never crossed my mind, but that's a very clever way to simulate remote instances without the pull mechanism

**Comment by tigattack** [comment]
I'd also really like to see a way to have a secondary Gatus instance push updates to a primary, as a sort of middle-ground between the existing 'remote instances' and 'external endpoints' features.

I have Gatus running on a VPS and would like to run an internal Gatus instance at home which can push endpoint statuses to my VPS instance. This would completely eliminate the need to expose my internal instance to the world or have a VPN tunnel between the two.

**Comment by iRazvan2745** [comment]
I would love to use this as an personal and enterprise solution at https://waffle.host
Looking forward to a stable release :)

**Comment by DeLoooping** [comment]
I'm currently experimenting with the ["remote instances"](https://github.com/TwiN/gatus?tab=readme-ov-file#remote-instances-experimental) feature and found that having an overview page with only remote endpoints does not work because gatus expects "at least 1 endpoint".

> gatus_overview-1  | goroutine 1 [running]:
> gatus_overview-1  | main.main()
> gatus_overview-1  | 	/app/main.go:31 +0x20f
> gatus_overview-1  | 2025/05/01 09:42:11 [main.configureLogging] Log Level is set to INFO
> gatus_overview-1  | 2025/05/01 09:42:11 [config.LoadConfiguration] Reading configuration from configFile=config/config.yaml
> gatus_overview-1  | panic: error parsing config: configuration should contain at least 1 endpoint

I'm not sure if this is the right place to report this issue, however.
Someone might move it if it's not :)

**Comment by ALParsons** [comment]
Here's an approach to consider:

An "exporter" (yaml node) feature that wraps your gatus endpoint status with a tee function that (in addition to updating the local instance) pushes status updates via event queue to an upstream instance. The upstream instance uses a "gatus importer" endpoint type that is specifically for receiving those events from a remote gatus. Maybe authenticate this with a PSK, or some other mutual auth that could be declared/rotated in a ci/cd pipeline. This should make sense later.

Each "gatus importer" nests its imported groups within a ui collection by importer id w/ a display name. In the main ui you might do this by duplicating the current element structure so it appears as a list of gatus instances on 1 page, or maybe you'd want pagination to support 1+n importers and have a header row that visualises the status of each importer. Maybe later down the track this could be a custom logo with a green/red health status indicator.

Then, you can imagine an architecture which can propagate each gatus instance through any kind of monitoring network you might imagine creating. Two gatus instances could export their endpoints and stream events to each other, or a chain of gatus instances could in effect stream their endpoint status events through a "tree" that can scale and bend to the shape of an organisation's requirements. Gatus could even be deployed headless inside k8s pods as a sidecar. The exporter defines the group, so many exporters can populate the same group (such as a group of k8s pods monitoring a given endpoint).

This setup requires no prior knowledge of an endpoint or group within the importer, only the exporter (i.e the auth key). A remote instance can create multiple exporters to the same importer (or multiple instances' exporters to the same importer) with each exporter joining its group inside the importer collection.

Each endpoint within each gatus has its own alerting config as it does today that is contained to that gatus instance, and a "gatus importer" may also have alerting parameters configured such as for an implied escalation (a pseudo call-tree). To configure this for a given group or endpoint within the importer would just require prior knowledge but that doesn't seem too bad as it would just match to any collection/group/endpoint if it exists.

I think this kind of architecture could enable heterogeneous gatus networks comprised of any mixture of one-to-many, one-to-one, many-to-one and many-to-many relationships. Within the yaml you may want to define importers and exporters in their own nodes to be referenced elsewhere, so you don't end up with 10's or 100's or more urls & configs across all the endpoints.

`importer:
    foo servers:
        id: foo-servers-us-east
        key: <secretkey>
    bar api:
        id: bar-king-mad
        key: <secretkey>

exporter:
    - foo servers:
        url: https://gatus-endpoint.public:3000
        id: foo-servers-us-east
        key: <secretkey>

endpoint:
... 
        exporter: foo servers`

**Comment by TomDeSchutter** [comment]
As requested to continue in this thread: I'm a huge fan of the remote instances feature and only just switched to the latest version and it's new UI. I combine multiple monitored sites into a single view, which worked perfect. To say, it still does, but:

Opening the remote instance by it's own website shows the full 50 blocks of endpoint history. On my main overview, all entries get listed but their history is limited to 20 blocks. 

Also, from the main overview, clicking through on and endpoint doesn't "resolve" if that makes sense. Just gets a blank page, but I've worked around by having buttons towards each seperate instance.

**Comment by TwiN** [comment]
@TomDeSchutter should be addressed by 7dac2cc

Good catch.

**Comment by eerison** [comment]
it is working perfect, thank you for this nice feature ❤️

**Comment by eerison** [comment]
I got an issue

I have 2 devices with the same name... on the first load it works fine! but after it fetch the data again, the remote device override the local one. and when I click on devices name, both goes to local's detail page.

Note: if you click on refresh icon on top, you won't see the issue, to reproduce change the automatic refresh from `5 min` to `10 seconds`

**Comment by TwiN** [comment]
@eerison it's not meant to merge endpoints results, only combine results from multiple instances into one endpoint.

In other words, try using a different name on your remote instance 😆

**Comment by green5-zero** [comment]
>     * `strategy: merge`: merge all statuses fetched from the remote instances with existing, overlapping service names, or create new statuses if they already exist
>       ![image](https://user-images.githubusercontent.com/15699766/106674764-e27d9680-6581-11eb-8b93-ad2d2395c05f.png)



This would be fantastic, and if possible some way to also configure notifications to wait for downtime from all or more than x instances before pushing anything out.

I think openstatus has some features for this but it's extremely hard to get working so it would be great if this could be supported, otherwise it's only closed source paid options. This is extremely helpful to avoid false emergency notifications if one status instance goes down or loses network.


If anyone wants to work on this I don't mind putting a bounty towards this to help accelerate development.

**Comment by surdaft** [comment]
Hi there! I have been tinkering with this for our production monitoring and noticed that the `client` configuration for remote seems to be undocumented.

[Config/Remote](https://github.com/TwiN/gatus/blob/bc433161d6067211e02cb27cb9870b781df6dc91/config/remote/remote.go) 
[Docs](https://github.com/TwiN/gatus#remote-instances-experimental)

I was encountering a timeout due to 2 instances living in alternative regions; APAC is on avg. 12s response time, US is on avg. 15s.

<img width="1590" height="502" alt="Image" src="https://github.com/user-attachments/assets/045125e3-8e87-4aed-9404-c7b015f4e17a" />

---

Also, due to the additional instances being loaded sequentially it caused the loading of the page to slow down. We could implement the use of goroutines but I feel that does not solve the problem all that well. Instead I think implementing a proxy endpoint for remote instances would allow the JS to promise the local instance and all remotes, then once they are all done we can handle display. This would mean they all load independently, the JS would handle the concurrency of the requests and the proxy means that the remote instances are kept securely behind the primary instance you are requesting through.

**Comment by lemeow125** [comment]
Hello!

Absolutely amazing work done with this project.

Is it possible to add native group prefixing for remote instances?

I'm working around this for now via my CI/CD setup: https://github.com/TwiN/gatus/discussions/1682

**Comment by selfhoster1312** [comment]
Hello, i'm new to the game here but i'm interested in this feature. I've started to make a few PRs to improve it:

- filter remote endpoints by group: #1712 
- fix detailed endpoint view from remote instance: #1715 
- allow a remote-only config: #1719 
- exclude remote endpoints from remote instance response (avoid semi-recursion / duplicates): #1721

I would really appreciate external opinions and review. I'm also happy to consolidate those different PRs into a single unified PR if that's more appropriate. If people want to test these PRs altogether i maintain [this branch](https://github.com/selfhoster1312/gatus/tree/fork) currently at commit db022588a9923611c3b292ce1ba437753bd3bfe8. EDIT: i should also mention i publish unsigned Debian packages for arm64/amd64 [here](https://github.com/ci-prebuild/gatus/releases) (please don't run untrusted binaries in production, but this may prove useful for testing)

I'm also noting from reading the discussions:

- using `instance.API` instead of `instance.URL` would make a cleaner code, but we should retain `instance.URL` not just for backwards-compatibility, but also for [magic retrieval of data from a non-gatus backend](https://github.com/TwiN/gatus/pull/330#discussion_r980572565)
- [sequential on-demand download can be slow](https://github.com/TwiN/gatus/issues/64#issuecomment-4288045760), it's worth exploring alternatives, such as adding an `instance.interval` and downloading remote data in the background (not on demand), maybe even automatically keeping track of remote instance health in the returned endpoints / in the UI (so you would still get a reply if a remote instance is timeouting, but indicating that requests to that remote failed)

**Comment by raffr85** [comment]
@green5-zero Are you still interested in funding the multi-instance alerting you described? My strongest programming experience is in Go, and I would be interested in a scoped implementation with a budget of US$1,000 or more, using disclosed Codex assistance.

I see that remote status aggregation and external endpoints already exist. The remaining requirement in your comment appears to be an alert decision based on multiple independent observers, including what happens when an observer itself loses connectivity. Before proposing a patch, could you confirm whether this is still needed, your funding range, and whether payment would depend on an upstream merge?

@TwiN Would that alerting scope be welcome in Gatus itself, or would you prefer it implemented outside the core? I have not started implementation or committed to a delivery date.
